Compilation error in result.hpp: 'FOut': is not a class or namespace name

After watching the Udemy course, I tried to integrate fplus in my projects, but I've encountered some errors, which I also encountered in an empty command-line project:
1>C:\Users\...\include\fplus\result.hpp(316,1): error C2653: 'FOut': is not a class or namespace name
1>C:\Users\...\include\fplus\result.hpp(323,1): error C2653: 'FOut': is not a class or namespace name

The code I tried is very simple:

`#include <iostream>`
`#include <fplus/fplus.hpp>`

`int main()`
`{`
`    std::cout << "Hello World!\n";`
`}`

My setup is Visual Studio version 16.5.3, compiler version set to ISO C++ 17 Standard

Ok, I've installed Visual Studio 2019 community in a VirtualBox.

I created a `main.cpp`:

```cpp
#include <fplus/fplus.hpp>
#include <iostream>

int main()
{
    std::list<std::string> things = {"same old", "same old"};
    if (fplus::all_the_same(things))
        std::cout << "All things being equal." << std::endl;
}
```

And using the command prompt I can compile without error:

```
cl.exe /I. /std:c++17 main.cpp
```

```
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28806 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

main.cpp
C:\Program Files\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.2880
1\include\ppltasks.h(1571): warning C4530: C++ exception handler used, but unwin
d semantics are not enabled. Specify /EHsc
Microsoft (R) Incremental Linker Version 14.26.28806.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:main.exe
main.obj
```

Could you please give me the exact command line (`cl.exe` callm generated by the IDE), that produces the compiler error?

@Dobiasd, i tried your repro, it gives error when compiler's [standard conformance mode](https://docs.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=vs-2019) switch `/permissive-` is provided:

```cmd
C:\git\FunctionalPlus\include> cl.exe /I. /permissive- /EHsc main.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28806 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

main.cpp
.\fplus/result.hpp(316): error C2653: 'FOut': is not a class or namespace name
.\fplus/result.hpp(323): error C2653: 'FOut': is not a class or namespace name
```

Looks like this upstream bug report needs some upvotes from community and love from VC compiler team: https://developercommunity.visualstudio.com/content/problem/675854/fmtlibs-safe-duration-cast-doesnt-compile-in-vs201.html. (I have cast mine) 😅

Meanwhile, this patch makes both compilation modes happy:

```diff
--- a/include/fplus/result.hpp
+++ b/include/fplus/result.hpp
@@ -304,14 +304,14 @@ auto compose_result(Callables&&... callables)
             internal::trigger_static_asserts<internal::check_arity_tag,
                                                  decltype(f),
                                                  decltype(args)...>();
-            using FOut = std::decay_t<
-                internal::invoke_result_t<decltype(f), decltype(args)...>>;

+            struct FOut : std::decay_t<
+                internal::invoke_result_t<decltype(f), decltype(args)...>> {};
             internal::trigger_static_asserts<internal::unary_function_tag,
                                                  decltype(g),
                                                  typename FOut::ok_t>();
-            using GOut = std::decay_t<
-                internal::invoke_result_t<decltype(g), typename FOut::ok_t>>;
+            struct GOut : std::decay_t<
+                internal::invoke_result_t<decltype(g), typename FOut::ok_t>> {};
             static_assert(std::is_same<typename FOut::error_t,
                                        typename GOut::error_t>::value,
                           "Error type must stay the same.");
```

@am11 That's perfect. Thanks a lot! With `/permissive-` I finally was able to reproduce the issue.

```
cl.exe /I. /std:c++17 /permissive- main.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28806 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

main.cpp
.\fplus/result.hpp(316): error C2653: 'FOut': is not a class or namespace name
.\fplus/result.hpp(323): error C2653: 'FOut': is not a class or namespace name
```

I also can confirm that your fix makes it work with and without `/permissive-`.

Would you like to submit it as a commit in a pull request?

--- a/include/fplus/result.hpp
+++ b/include/fplus/result.hpp
@@ -304,14 +304,24 @@ auto compose_result(Callables&&... callables)
             internal::trigger_static_asserts<internal::check_arity_tag,
                                                  decltype(f),
                                                  decltype(args)...>();
+#if _MSC_VER >= 1920 // in VS2019, compilation with /permissive- breaks with 'using' syntax below
+            struct FOut : std::decay_t<
+                internal::invoke_result_t<decltype(f), decltype(args)...>> {};
+#else
             using FOut = std::decay_t<
                 internal::invoke_result_t<decltype(f), decltype(args)...>>;
+#endif
 
             internal::trigger_static_asserts<internal::unary_function_tag,
                                                  decltype(g),
                                                  typename FOut::ok_t>();
+#if _MSC_VER >= 1920 // in VS2019, compilation with /permissive- breaks with 'using' syntax below
+            struct GOut : std::decay_t<
+                internal::invoke_result_t<decltype(g), typename FOut::ok_t>> {};
+#else
             using GOut = std::decay_t<
                 internal::invoke_result_t<decltype(g), typename FOut::ok_t>>;
+#endif
             static_assert(std::is_same<typename FOut::error_t,
                                        typename GOut::error_t>::value,
                           "Error type must stay the same.");

